### PR TITLE
Refactor integration test suite to use testify

### PR DESCRIPTION
**Pull request by DharmitD**
## The issue resolved by this Pull Request:
Resolves https://issues.redhat.com/browse/RHOAIENG-1672

## Description of your changes:

Convert integration tests to leverage testify instead of gomega/ginkgo.

## Testing instructions

Deploy DSPO. In a fresh namespace, run the following command:

```
go test ./... --tags=test_integration -v -kubeconfig=${KUBECONFIG_PATH} -k8sApiServerHost=${TARGET_CLUSTER}   -DSPANamespace=${TARGET_NAMESPACE} -DSPAPath=resources/dspa-lite.yaml
```

For the impatient developer, you can use the following flag to skip DSPA install. This is useful when you want to make changes to a live environment and run the tests against it:
```
go test ./... --tags=test_integration -v -kubeconfig=${KUBECONFIG_PATH} -k8sApiServerHost=${TARGET_CLUSTER}   -DSPANamespace=${TARGET_NAMESPACE} -DSPAPath=resources/dspa-lite.yaml -skipDeploy=true -skipCleanup=true
```

## Checklist
- [x] The commits are squashed in a cohesive manner and have meaningful messages.
- [x] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [x] The developer has manually tested the changes and verified that the changes work

### CODE REVIEW

**Comment by DharmitD** [comment]
> Worked perfectly fine for complete integration test but the test case where we skip the DSPA install using the flag fails as it expects a DSPA named `test-dspa`. Is that a requirement?
> 
> Full Test Case
> ```
> go test ./... --tags=test_integration -v -kubeconfig=${KUBECONFIG_PATH} -k8sApiServerHost=${TARGET_CLUSTER}   -DSPANamespace=${TARGET_NAMESPACE} -DSPAPath=resources/dspa-lite.yaml
> ?   	github.com/opendatahub-io/data-science-pipelines-operator	[no test files]
> ?   	github.com/opendatahub-io/data-science-pipelines-operator/api/v1alpha1	[no test files]
> ?   	github.com/opendatahub-io/data-science-pipelines-operator/controllers	[no test files]
> ?   	github.com/opendatahub-io/data-science-pipelines-operator/controllers/config	[no test files]
> ?   	github.com/opendatahub-io/data-science-pipelines-operator/controllers/testutil	[no test files]
> ?   	github.com/opendatahub-io/data-science-pipelines-operator/controllers/util	[no test files]
> ?   	github.com/opendatahub-io/data-science-pipelines-operator/tests/util	[no test files]
> === RUN   TestIntegrationTestSuite
> SetupSuite started
> 2024-03-12T23:49:09+05:30	INFO	Deploying DSPA...
> 2024-03-12T23:49:10+05:30	INFO	Waiting for DSPA pods to ready...
> 2024-03-12T23:50:40+05:30	INFO	DSPA Deployed.
> 2024-03-12T23:50:40+05:30	INFO	Setting up Portforwarding service.
> Forwarding service: ds-pipeline-test-dspa to pod ds-pipeline-test-dspa-77d8ff84cd-nmp25 ...
> Forwarding from 127.0.0.1:8888 -> 8888
> Forwarding from [::1]:8888 -> 8888
> 2024-03-12T23:50:42+05:30	INFO	Portforwarding service Successfully set up.
> SetupSuite completed
> === RUN   TestIntegrationTestSuite/TestAPIServerDeployment
> === RUN   TestIntegrationTestSuite/TestAPIServerDeployment/Should_successfully_fetch_pipelines
> Handling connection for 8888
> 2024-03-12T23:50:43+05:30	INFO	{"pipelines":[{"pipeline_id":"a4c0a3aa-c953-4b21-80cd-fabf0629a898","display_name":"[Demo] iris-training","description":"[source code](https://github.com/opendatahub-io/data-science-pipelines/tree/master/samples/iris-sklearn) A simple pipeline to demonstrate a basic ML Training workflow","created_at":"2024-03-12T18:20:32Z"}],"total_size":1}
> === RUN   TestIntegrationTestSuite/TestAPIServerDeployment/Should_successfully_upload_a_pipeline
> 2024-03-12T23:50:44+05:30	INFO	{"pipeline_id":"74c97e0f-fb39-4d1a-9a5a-ac239fabab3f","display_name":"test-pipeline-run.yaml","created_at":"2024-03-12T18:20:43Z"}
> === RUN   TestIntegrationTestSuite/TestDSPADeployment
> === RUN   TestIntegrationTestSuite/TestDSPADeployment/with_default_MariaDB_and_Minio
> === RUN   TestIntegrationTestSuite/TestDSPADeployment/with_default_MariaDB_and_Minio/should_have_8_pods
> === RUN   TestIntegrationTestSuite/TestDSPADeployment/with_default_MariaDB_and_Minio/should_have_a_ready_ds-pipeline-test-dspa_deployment
> === RUN   TestIntegrationTestSuite/TestDSPADeployment/with_default_MariaDB_and_Minio/should_have_a_ready_ds-pipeline-persistenceagent-test-dspa_deployment
> === RUN   TestIntegrationTestSuite/TestDSPADeployment/with_default_MariaDB_and_Minio/should_have_a_ready_ds-pipeline-scheduledworkflow-test-dspa_deployment
> === RUN   TestIntegrationTestSuite/TestDSPADeployment/with_default_MariaDB_and_Minio/should_have_a_ready_mariadb-test-dspa_deployment
> === RUN   TestIntegrationTestSuite/TestDSPADeployment/with_default_MariaDB_and_Minio/should_have_a_ready_minio-test-dspa_deployment
> === RUN   TestIntegrationTestSuite/TestFetchExperiments
> === RUN   TestIntegrationTestSuite/TestFetchExperiments/Should_successfully_fetch_experiments
> 2024-03-12T23:50:47+05:30	INFO	{"experiments":[{"experiment_id":"125e7b72-4ddc-4cd9-b17f-bc93a548ffa8","display_name":"Default","description":"All runs created without specifying an experiment will be grouped here.","created_at":"2024-03-12T18:20:34Z","storage_state":"AVAILABLE"}],"total_size":1}
> --- PASS: TestIntegrationTestSuite (100.86s)
>     --- PASS: TestIntegrationTestSuite/TestAPIServerDeployment (1.82s)
>         --- PASS: TestIntegrationTestSuite/TestAPIServerDeployment/Should_successfully_fetch_pipelines (0.82s)
>         --- PASS: TestIntegrationTestSuite/TestAPIServerDeployment/Should_successfully_upload_a_pipeline (1.00s)
>     --- PASS: TestIntegrationTestSuite/TestDSPADeployment (2.78s)
>         --- PASS: TestIntegrationTestSuite/TestDSPADeployment/with_default_MariaDB_and_Minio (2.78s)
>             --- PASS: TestIntegrationTestSuite/TestDSPADeployment/with_default_MariaDB_and_Minio/should_have_8_pods (1.15s)
>             --- PASS: TestIntegrationTestSuite/TestDSPADeployment/with_default_MariaDB_and_Minio/should_have_a_ready_ds-pipeline-test-dspa_deployment (0.54s)
>             --- PASS: TestIntegrationTestSuite/TestDSPADeployment/with_default_MariaDB_and_Minio/should_have_a_ready_ds-pipeline-persistenceagent-test-dspa_deployment (0.28s)
>             --- PASS: TestIntegrationTestSuite/TestDSPADeployment/with_default_MariaDB_and_Minio/should_have_a_ready_ds-pipeline-scheduledworkflow-test-dspa_deployment (0.27s)
>             --- PASS: TestIntegrationTestSuite/TestDSPADeployment/with_default_MariaDB_and_Minio/should_have_a_ready_mariadb-test-dspa_deployment (0.27s)
>             --- PASS: TestIntegrationTestSuite/TestDSPADeployment/with_default_MariaDB_and_Minio/should_have_a_ready_minio-test-dspa_deployment (0.27s)
>     --- PASS: TestIntegrationTestSuite/TestFetchExperiments (0.72s)
>         --- PASS: TestIntegrationTestSuite/TestFetchExperiments/Should_successfully_fetch_experiments (0.72s)
> PASS
> ok  	github.com/opendatahub-io/data-science-pipelines-operator/tests	100.873s
> ```
> 
> Test case skipping the DSPA install with the flag

Yep, this is expected. Basically the skip deploy command works when you have a DSPA instance already up with a DSP CR named "test-dspa". Hence it would fail on not finding that CR.

**Review by amadhusu — APPROVED** [review]

**Comment by HumairAK** [comment]
@DharmitD can you either edit the message to be more clear or squash this commit "[fixes](https://github.com/opendatahub-io/data-science-pipelines-operator/pull/589/commits/b0677ae376284f2d0b036ca1c90953f43f1644cb)"

**Comment by HumairAK** [comment]
@diegolovison should review this as well, as he'll need to update the test script that runs downstream

File: tests/suite_test.go [hunk 1]
```diff
@@ -1,35 +1,23 @@
 //go:build test_integration
 
-/*
```
Comment (HumairAK) [inline tests/suite_test.go:3]: can you re-add all the apache license notices where they have been removed?:

File: tests/suite_test.go [hunk 2]
```diff
@@ -90,19 +73,27 @@ type ClientManager struct {
...
-
-	// Inform Ginkgo to start the test suite, passing it the *testing.T instance and a description of the suite.
-	// Only call RunSpecs once and let Ginkgo worry about calling *testing.T for us.
-	RunSpecs(t, "Controller Suite")
+type IntegrationTestSuite struct {
+	suite.Suite
+	Clientmgr     ClientManager                                 // Note the capitalization to export
+	Ctx           context.Context                               // Exported and added
+	DSPANamespace string                                        // Already exported, make sure it's part of the struct
+	DSPA          *dspav1alpha1.DataSciencePipelinesApplication // Exported and added
```
Comment (HumairAK) [inline tests/suite_test.go:81]: can you explain these comments? exported and added where?

File: tests/suite_test.go [hunk 3]
```diff
@@ -113,11 +104,8 @@ func init() {
 	flag.StringVar(&DSPANamespace, "DSPANamespace", DefaultDSPANamespace, "The namespace to deploy DSPA.")
 
 	flag.DurationVar(&DeployTimeout, "DeployTimeout", DefaultDeployTimeout, "Seconds to wait for deployments. Consider increasing this on resource starved environments.")
-	DeployTimeout *= time.Second
 	flag.DurationVar(&PollInterval, "PollInterval", DefaultPollInterval, "Seconds to wait before retrying fetches to the api server.")
-	PollInterval *= time.Second
 	flag.DurationVar(&DeleteTimeout, "DeleteTimeout", DefaultDeleteTimeout, "Seconds to wait for deployment deletions. Consider increasing this on resource starved environments.")
```
Comment (HumairAK) [inline tests/suite_test.go:108]: you set the default time here: https://github.com/opendatahub-io/data-science-pipelines-operator/pull/589/files#diff-96e66e1ccb0083ac5a2b7fc37cb002803d242ea67c1981a5d09ef8ee2668354cR61-R63 

but it looks like if the user specifies a time, we don't convert to the proper time.duration, did you confirm this? it makes sense to me to keep this here, but remove it from the const()

File: tests/suite_test.go [hunk 4]
```diff
@@ -126,81 +114,81 @@ func init() {
 	flag.BoolVar(&skipCleanup, "skipCleanup", DefaultSkipCleanup, "Skip DSPA cleanup.")
 }
 
-var _ = BeforeSuite(func() {
-	ctx, cancel = context.WithCancel(context.TODO())
+func (suite *IntegrationTestSuite) SetupSuite() {
+	fmt.Println("SetupSuite started") // Debug statement
```
Comment (HumairAK) [inline tests/suite_test.go:118]: // Debug statement 
Did you mean to remove this?

File: tests/suite_test.go [hunk 5]
```diff
@@ -126,81 +114,81 @@ func init() {
...
+	forwarderResult, err = forwarder.WithForwarders(ctx, options, kubeconfig)
+	suite.Require().NoError(err)
 	_, err = forwarderResult.Ready()
-	Expect(err).NotTo(HaveOccurred())
-	loggr.Info("Portforwarding service Successfully set up.")
+	suite.Require().NoError(err)
 
 	APIServerURL = fmt.Sprintf("http://127.0.0.1:%d", PortforwardLocalPort)
+	loggr.Info("Portforwarding service Successfully set up.")
+	fmt.Println("SetupSuite completed") // Debug statement
```
Comment (HumairAK) [inline tests/suite_test.go:178]: > // Debug statement

did you mean to remove this?

File: tests/experiments_test.go [hunk 6]
```diff
@@ -20,21 +20,24 @@ package integration
...
-		loggr.Info(string(responseData))
+		require.NoError(t, err, "Error fetching experiments")
 
-		Expect(err).ToNot(HaveOccurred())
-		Expect(response.StatusCode).To(Equal(200))
+		responseData, err := ioutil.ReadAll(response.Body)
+		defer response.Body.Close()
+		require.NoError(t, err, "Error reading response body")
 
+		// Corrected to use suite assertion methods correctly
```
Comment (HumairAK) [inline tests/experiments_test.go:39]: remove this

File: tests/experiments_test.go [hunk 7]
```diff
@@ -20,21 +20,24 @@ package integration
...
+		suite.Assert().Equal(200, response.StatusCode, "Expected HTTP status code 200 for fetching experiments")
+		loggr.Info(string(responseData)) // Assuming 'loggr' is globally accessible
```
Comment (HumairAK) [inline tests/experiments_test.go:41]:
```suggestion
		loggr.Info(string(responseData)) 
```

File: tests/suite_test.go (hunk 2 above)
Comment (DharmitD) [reply tests/suite_test.go]: Removed these comments, not needed anymore.

File: tests/suite_test.go (hunk 4 above)
Comment (DharmitD) [reply tests/suite_test.go]: Yes, removed the debug statements, just added them to test locally.

File: tests/suite_test.go (hunk 5 above)
Comment (DharmitD) [reply tests/suite_test.go]: Yes, removed the print line.

File: tests/suite_test.go (hunk 3 above)
Comment (DharmitD) [reply tests/suite_test.go:126]: Done, kept it here and removed from const()

File: Makefile [hunk 8]
```diff
@@ -128,7 +128,7 @@ functest: manifests generate fmt vet envtest ## Run tests.
 .PHONY: integrationtest
 integrationtest: ## Run integration tests
 	cd tests && \
-	go run github.com/onsi/ginkgo/v2/ginkgo --tags=test_integration -- -kubeconfig=${KUBECONFIGPATH} -k8sApiServerHost=${K8SAPISERVERHOST} -DSPANamespace=${DSPANAMESPACE} -DSPAPath=${DSPAPATH} -ginkgo.v
+	go test ./... --tags=test_integration -v -kubeconfig=${KUBECONFIGPATH} -k8sApiServerHost=${K8SAPISERVERHOST} -DSPANamespace=${DSPANAMESPACE} -DSPAPath=${DSPAPATH}
```
Comment (gmfrasca) [inline Makefile:131]: not sure we should be using verbose mode here

Comment (DharmitD) [reply Makefile:131]: Isn't `-gingko.v` in the previous command also verbose mode though? Which is why I set this command to verbose IIRC.

Comment (gmfrasca) [reply Makefile:131]: hmm so unittest has -v but functest doesn't.  we should converge on one but for now this is fine i suppose

**Review by gmfrasca — APPROVED** [review]

**Pull request merged**